# Assignment 4: Keyphrase Extraction, Named Entity Recognition & Neural Models

Due: Monday, February 06, 2023, at 2pm via Moodle

**Team Members** `<Fill out>`

Please note that this assignment comes with quite a number of artifacts, totaling somewhere around 5 GB of necessary disk space. In case you are running into issues or do want to keep your environment "clean", we suggest the use of [Google Colab](https://colab.research.google.com/).

In [13]:
%%bash
. ~/.bashrc
python3 -m pip install keybert
python3 -m pip install git+https://github.com/LIAAD/yake
python3 -m pip install transformers
python3 -m pip install datasets
python3 -m pip install nltk
python3 -m pip install spacy
# Install necessary packages for all questions

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/LIAAD/yake to /tmp/pip-req-build-eh5cpb57
  Resolved https://github.com/LIAAD/yake to commit 8d71d94ded93fb77f1361f62e5264f19b9c91cd7
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


  Running command git clone --filter=blob:none --quiet https://github.com/LIAAD/yake /tmp/pip-req-build-eh5cpb57
ERROR: Could not find a version that satisfies the requirement urllib2 (from versions: none)
ERROR: No matching distribution found for urllib2


CalledProcessError: ignored

## Task 1: Keyphrase Extraction (5 + 3 + 3 + 5) = 16 Points

In this task, we will implement our own unsupervised keyphrase extraction (KPE) module utilizing a simple grammatical ruling system, which we apply to a Sherlock Holmes novel.
To generate TF-IDF-weighted phrases, we will be using the entire collection of Sir Arthur Donan Coyle novels to calculate document frequencies.

Finally, we compare the results to general-purpose KPE libraries.

### Sub Task 1: Unsupervised Keyphrase Extraction System (5 Points)

#### 1. Candidate Generation
We will need to generate a set of suitable candidate phrases first, which can then be ranked as keyphrases later on. To do this, we will again be using spaCy's, this time its rule-based [`Matcher` class](https://spacy.io/api/matcher).

The syntactic pattern of a keyphrase candidate should satisfy the following rules:

1. An optional adjective, noun, proper noun
2. An optional adjective, noun, proper noun
3. A mandatory noun or proper noun.

Add a second pattern, which recognizes the pattern

1. A noun or proper noun
2. An adposition
3. Another noun or proper noun

Note that the first condition will match any phrase of length between 1-3 tokens, which is a suitable approximation for our task at hand, whereas the second pattern is slightly more specific, always matching exactly three tokens.
An example of a valid matched phrases for the first pattern would be "Sherlock Holmes" ([PROPN, PROPN]), and "Hounds of Baskervilles" ([NOUN, ADP, PROPN]) for the second pattern.

In [101]:
import spacy
from spacy.matcher import Matcher
from urllib.request import urlopen
from collections import Counter
from yake import KeywordExtractor


In [86]:
# load language model
nlp = spacy.load("en_core_web_sm", disable=["ner"])
matcher = Matcher(nlp.vocab)

# Define the above patterns
patterns = [
            [{"POS": {"IN": ["ADJ", "NOUN","PROPN"]}, "OP": "*"},
             {"POS": {"IN": ["ADJ", "NOUN","PROPN"]}, "OP": "*"},
             {"POS": {"IN": ["NOUN","PROPN"]}}],
            
            [{"POS": {"IN": ["NOUN","PROPN"]}},
             {"POS": "ADP"},
             {"POS": {"IN": ["NOUN","PROPN"]}}]
            ]


matcher.add("Sherlock", patterns)

To verify whether your pattern is correct, use the below example.
If you have done everything correctly, your matcher will identify **13 phrases**.

In [87]:
doc = nlp("This is a simple test. It should return 'simple', and 'test', among other phrases. Maybe we can also see if it can recognize the art of war. Would it recognize integer linear programming, too?")
matches = matcher(doc)

print(len(matches))

13


#### 2. Applying Your System

Once you have matched the correct number of keyphrase candidates on the above example, apply your rule-based matcher to an actual data sample. We are going to use the Sherlock Holmes novel "Hounds of Baskervilles". You can find the raw text file at the following URL:

https://sherlock-holm.es/stories/plain-text/houn.txt

Download the text from this URL and apply your spaCy model and matcher on it.  
**Hint:** Make sure you properly decode your input, since some libraries return binary strings.

In [88]:
from urllib.request import urlopen
def load_txt_from_url(url: str = "https://sherlock-holm.es/stories/plain-text/houn.txt") -> str:
  text = ""
  for line in urlopen(url):
    line = line.decode('UTF-8')
    text += str(line)
  return text

text = load_txt_from_url()

# Apply the spacy model to the loaded text and extract the phrases with the Matcher
doc = nlp(text)
matches = matcher(doc)
print(len(matches))

15938


We will now investigate which phrase candidates are the most frequently appearing in this novel, simply based on the phrase frequency. Therefore, convert your abstract match objects into actual strings, lowercase them, and return the 20 most frequently occurring phrase candidates and their respective frequencies.  
**Hint:** For counting occurrences, you may look at `collections.Counter`.

In [89]:
candidates = []
# Lowercase and add the extracted candidate matches to `candidates`
for match_id,start,end in matches:
    string = str(doc[start:end])
    string = string.lower()
    candidates.append(string)

# Count the number of occurrences of different candidate phrases
print(type(candidates))
candidate_phrases = Counter(candidates)
# Print the most frequently occurring phrases, together with the respective frequencies
print(candidate_phrases.most_common(20))

<class 'list'>
[('sir', 350), ('man', 214), ('holmes', 192), ('moor', 159), ('henry', 156), ('sir henry', 135), ('watson', 117), ('baskerville', 116), ('dr.', 109), ('charles', 94), ('stapleton', 93), ('mortimer', 89), ('night', 88), ('time', 86), ('sir charles', 86), ('house', 75), ('face', 75), ('hound', 72), ('barrymore', 72), ('eyes', 71)]


#### 3. Briefly summarize the quality of your top 20 candidates:

The Top 20 Candidates are not really sentences, but mostly single words, as the pattern from previously allows amtches of single tokens, that are much more common, than token combinations. This is not really what we wanted. 

### Sub Task 2: Generating Document Frequency Values (3 Points)

To compare the previously generated terms with a more refined model, we are going to extract document frequencies from the collection of all Sherlock Holmes works. Since the books are relatively long documents, we are instead going to split based on a simple heuristic in the input document, which should allow a decent approximation by taking into account individual chapters of each novel.

1. Start by loading the Sherlock Holmes canon from https://sherlock-holm.es/stories/plain-text/cnus.txt  
Afterwards, split the full document into individual chapters. For this, use three consecutive line breaks `\n\n\n` as a splitting condition to approximate the chapters.

In [90]:
df_texts = load_txt_from_url("https://sherlock-holm.es/stories/plain-text/cnus.txt")

split_df_texts = df_texts.split("\n\n\n")

print(len(split_df_texts))

353


After splitting, you should have 353 individual "documents" to work with.

2. Now, create a dictionary containing each phrase encountered in the larger corpus, and its associated document frequency. Again, ensure that phrase strings are lowercased for consistency with the previous transformation.  
**Hint:** Since the processing of 353 documents might take a while, incorporate [`tqdm.tqdm`](https://tqdm.github.io/) to visualize progress on the task.

In [91]:
from tqdm import tqdm
from typing import List

def return_occurring_phrases(doc_text: str) -> List[str]:
  # process text with spaCy and apply the Matcher
  doc = nlp(doc_text)
  matches = matcher(doc)

  # Candidates can be a set, since we only care about the occurrence *once* for IDF values.
  # Again, extract the lower-cased text of a matched span.
  candidates = set()
  for match_id, start, end in matches:
    string = str(doc[start:end])
    string = string.lower()
    candidates.add(string)

  return list(candidates)

all_document_phrases = []
# Iterate through the individual documents and extract phrases for them. Use `tqdm` to visualize progress
for element in tqdm(split_df_texts):
  candidates = return_occurring_phrases(element)
  for candidate in candidates:
    all_document_phrases.append(candidate)



# Once again, count the frequency of term occurrences across all documents
phrases = Counter(all_document_phrases)
# Print the most frequently occurring phrases, together with the respective frequencies




100%|██████████| 353/353 [01:09<00:00,  5.05it/s]


3. Output the 20 most frequently appearing document phrases that your system detected:

In [92]:
print(phrases.most_common(20))

[('man', 112), ('holmes', 107), ('eyes', 104), ('time', 104), ('night', 104), ('face', 102), ('hand', 102), ('sherlock', 101), ('sherlock holmes', 101), ('way', 101), ('matter', 100), ('room', 99), ('mr.', 98), ('day', 97), ('hands', 97), ('case', 96), ('house', 96), ('life', 96), ('one', 95), ('door', 95)]


### Sub Task 3: Generating Weighted Keyphrases (3 Points)

We can now incorporate the extracted keyphrases to calculate `tf-idf` scores, and return a hopefully improved version of our keyphrases for the original "Hounds of Baskervilles" document. 

1. Iterate over all phrases occurring in the novel "Hounds of Baskervilles", and re-score phrases according to the definition of TF-IDF. Use the smoothed definition of idf:

$ idf(t, D) = \log \frac{|D|}{|\{d \in D : t \in d\}| + 1} + 1 $

In [93]:
import math
from typing import Dict

def tf_idf(tf: int, df_count: int) -> float:
  """
  Computes the TF-IDF scores according to the above-mentioned formula.
  Note that you may use a constant for the number of documents (|D|).
  """
  idf = math.log(1/1+df_count) + 1

  return tf*idf

tf_idf_weighted_candidates = []

# Iterate through all candidate phrase/frequency pairs and compute the TF-IDF scores for each phrase
# Store the phrase together with its TF-IDF score in `tf_idf_weighted_candidates`

for candidate, tf in candidate_phrases.items():
  df_count = phrases[candidate]
  tfidf = tf_idf(tf,df_count)
  helpertuple = (candidate, tfidf)
  tf_idf_weighted_candidates.append(helpertuple)

2. Now print the top 20 candidate phrases by TF-IDF weight, and compare the results to your previous output. 

In [99]:
print(sorted(tf_idf_weighted_candidates,key=lambda a: a[1], reverse = True)[:20])

[('sir', 1917.0678850673723), ('man', 1225.660993204441), ('holmes', 1090.9691956078502), ('watson', 640.8484072939502), ('henry', 638.2026227238973), ('moor', 627.165797687464), ('dr.', 553.4515813857234), ('sir henry', 509.29947750237045), ('night', 497.54851081386204), ('time', 486.240590113547), ('baskerville', 451.28312391595506), ('face', 422.6046741172227), ('house', 418.10332338775373), ('eyes', 401.43118486118414), ('charles', 384.5579906156817), ('way', 359.9982600501933), ('hall', 358.8994069531564), ('mr.', 352.49255055847914), ('day', 351.85295115624604), ('case', 351.2067916457131)]


3. Write your insights on the comparison of the results below. Try to theorize why some of the phrases still appear, or why other phrases are no longer present:

The lists are very similar in the way, that most values from the first list are reappearing. A notable example of an element that is now missing, is "Mortimer". A new element, that previously was not present, is "life" ore "one". A reason for this could be, that we are using the term frequency from various different novels. Maybe the character named "Mortimer" does only play a significant role in the Hound of Baskerville, that we used for the first example, therefore he has no high document frequency which is why he might have dropped out of the list. Words like "life" or "one" are more general and tend to be present across multiple documents whicht could explain why we find them here. 

4. Give two examples of how you could further improve the list of keyphrase values.

1. We could increase the body of work, that the keyphreases are based on, e.g. using more novels of different authors to gain a better understanding of the matter at hand.
2. We could filter out stop words to eliminate single tokens like "sir" if they are not bound to a second element. Currently we have "sir" as well as "sir henry" which is not optimal



### Sub Task 4: Apply off-the-shelf Keyphrase Extraction Tools (5 Points)

To put the findings of your system into context, compare them with two popular open-source libraries, namely [YAKE!](https://github.com/LIAAD/yake) and [KeyBERT](https://github.com/MaartenGr/KeyBERT).

1. First, start by running the document with YAKE!; you may use the default parameters. Print the resulting keyphrases, which by default returns 20 phrases.

In [133]:

extractor = KeywordExtractor()
keywords = extractor.extract_keywords(text)

# Print the top 20 keywords
print(keywords)

[('Sir Henry Baskerville', 9.703823247628816e-06), ('Sir Henry', 1.648100191408312e-05), ('Sir Charles Baskerville', 2.0351834688440424e-05), ('Sir Charles', 3.724150731612377e-05), ('Sir', 0.00010788043646785337), ('Sir Charles death', 0.0001335320154208488), ('Henry Baskerville', 0.00020407970787866275), ('Holmes', 0.00027267641332312374), ('Hall Sir Henry', 0.0002827705937659273), ('Baskerville Hall', 0.00030297998752354974), ('Sherlock Holmes', 0.00030366832269808347), ('friend Sir Henry', 0.0003232815192427155), ('Baskerville Hall Sir', 0.00033497855089080193), ('Henry', 0.0003734244162998363), ('Charles Baskerville', 0.00045232284621121374), ('BASKERVILLES Arthur Conan', 0.00046235145087207454), ('asked Sir Henry', 0.0005001869412629496), ('Sir Henry put', 0.0005065910358925609), ('Arthur Conan Doyle', 0.0005408670117356265), ('Baskerville', 0.0005977236792600148)]


2. Compare both runtime efficiency and the extracted phrases with your own system.

The of the shelf solution runs in 6 seconds while our system needs about a Minute of time. Therefore the off the shelf solutions seems to run much more efficiently. The extracted solutions are also totally different from our system. The off the shelf solution extracts only named entities, ignoring the words like night or face, that our solution held, completely. Also the off the shelf solution shares the same problems as ours when it comes to doubles. It also extracts "Sir Henry Baskerville" as well as "Sir Henry" which is not optimal, as the words stem from the same string. 


3. Now use the KeyBERT library to extract keyphrases. Importantly, you will need to split the document into separate paragraphs, as the underlying neural model will be unable to handle the complete document as input.  
Use the pattern of `\n\n` to separate the text into smaller paragraphs, and filter out any empty lines after. An "empty line" also constitutes all inputs that only contain newline (`\n`) or whitespace ` ` characters.


In [106]:
# Split the input text according to the specified criteria and filter empty lines out.
text = load_txt_from_url()
split_text = text.split("\n\n")
print(len(split_text))

bad_text = (""," ","\n")

for element in split_text:
  if element in bad_text:
    split_text.remove(element)

print(len(split_text))




1552
1527


4. To ensure consistency between the tools when extracting keyphrases, set the *n*-gram range to `(1,3)`.
Otherwise, leave all parameters at the default value, and extract the keyphrases from each paragraph.

In [134]:
from keybert import KeyBERT

# This might take a while to install
model = KeyBERT("all-MiniLM-L6-v2")

# Extract the keyphrases from each split, using the adjusted keyphrase ngram range
# Hint: You may pass a list to the extraction function and KeyBERT will automatically handle iteration.
extracted_phrases = model.extract_keywords(split_text, keyphrase_ngram_range=(1, 3))

5. Combine the predictions of all individual splits into a single list. For this, sum up the prediction scores across all splits.  
**Hint:** `collections.defaultdict` makes aggregations like this much easier.

In [131]:
from typing import List, Tuple
from collections import defaultdict

def merge_predictions(list_of_predictions: List[List[Tuple]]) -> List[Tuple]:
    """
    Combines lists of predictions into a single list with added scores.
    """
    listelements = []

    for liste in list_of_predictions:
      for element in liste:
          listelements.append(element)

    # Initialisation of defaultdict
    phrase_dict = defaultdict(int)
    
    for k, v in listelements:
        phrase_dict[k] += v


    # Extract the 20 keyphrases with the highest weithgts from `phrase_dict`
    #phrase_list = sorted(phrase_dict,key=lambda a: a[1], reverse = False)[:20]
    phrase_list = Counter(phrase_dict).most_common(20)

    return phrase_list

In [132]:
print(merge_predictions(extracted_phrases))

[('sir henry', 30.826300000000003), ('holmes', 30.425000000000004), ('dr mortimer', 24.266699999999997), ('mortimer', 22.4), ('watson', 21.2683), ('sir charles', 17.5742), ('said holmes', 16.1206), ('sherlock holmes', 15.2558), ('barrymore', 13.793199999999999), ('dr watson', 12.346900000000002), ('mr holmes', 10.4943), ('sir', 10.3469), ('moor', 10.3131), ('mr sherlock holmes', 8.103), ('hound', 7.942399999999999), ('sir henry baskerville', 7.7776), ('henry', 7.257099999999999), ('baskerville hall', 5.9735000000000005), ('sherlock', 5.7341999999999995), ('yes', 5.649700000000001)]


6. Again, evaluate the result and compare it to the other two approaches in terms of extraction quality and extraction speed.

The extraction of the Keyphrases with KeyBERT took about 2 Minutes and 42 Seconds. This makes it the slowest of all the solutions, even compared to our own solution. It results in a very different list of keyphrases than the other solutions. While the results are mostly named entities, this time there are also some different phrases like "yes" or "moor". While the results seem to be better, than the ones we got with YAKE, KeyBERT has some of the same problems with doubles as the other two approaches. For example, "watson" and "dr. watson" both show up in the results. It would have been better, if KeyBERT had recognized these two results as stemming from the same family of strings, merging them into one. While it is less efficient than the other approaches, KeyBERT delivers the best results, as ist has less problems with duplicates than YAKE without relying on the text corpus from the other sherlock holmes novels. This results in "DR Mortimer" being in the results again which speaks for the quality of the method. 

## 2. Named Entity Recognition (4 + 5 + 5 = 14 Points)

Slightly different, but still operating on the sequence level, is the task of Named Entity Recognition (NER).
In this task, we will evaluate the NER capabilities of some more open-source libraries.
Particularly, we will also evaluate the utility of NER as a stand-in for Keyphrase Extraction.

### Sub Task 1: Using spaCy NER (4 Points)

So far, when using spaCy models, we have primarily disabled the NER component, as it requires significant extra compute.
In this task, we will explicitly leave the component enabled, to see what results it can produce on the text from the previous question.

In [ ]:
import spacy

# Load the en_core_web_sm model, but with NER enabled.
nlp = spacy.load("en_core_web_sm")

1. Re-load the text for the "Hounds of Baskervilles" novel, and run it with the spacy model.

In [ ]:
# Re-use the function from the previous exercise.
text = ## YOUR CODE

doc = ## YOUR CODE

2. Similar to the previous exercise, count the number of occurrences, however, this time for the extracted entities instead of phrases. Print the top 20 most frequently occurring entities.  
Make sure to lowercase the text again during your aggregation.

In [ ]:
# Count the number of occurrences of particular entities
## YOUR CODE

# Print the top 20 most frequently occurring entities.
print( ## YOUR CODE

You might have noticed some unwanted results in the list, such as "night". Upon closer inspection, it turns out that the NER module further differentiates between different entity *categories*, such as PERSON (referencing, as expected, a physical person) or ORG (organizations, such as companies, NGOs, etc.), but also TIME (under which "night" falls). For reference, you can find the full list of supported NER labels by this particular model [here](https://spacy.io/models/en#en_core_web_sm-labels).

3. Refine the list of most common entities by printing out the top three occurring entities in the category `PERSON`, `ORG` and `GPE` (physical locations) instead.

In [ ]:
def get_top_entities_by_class(doc: spacy.tokens.Doc, class_name: str, n: int = 3) -> List[Tuple[str, int]]:
    """
    Returns the three most frequent entities (and their frequencies)
    of entity type `class_name` from `doc`.
    """
    # Extract phrase and frequency of a particular entity class
    counter = ## YOUR CODE
    # Return the top 3 entities and frequencies
    return ## YOUR CODE

# Print the results for "PERSON", "ORG" and "GPE"
print( ## YOUR CODE
print( ## YOUR CODE
print( ## YOUR CODE

### Sub Task 2: Financial Bank Statements of Deutsche Bank (5 Points)

Instead of using the Sherlock Holmes Novels, we will now compare the functionality of spaCy and NLTK's NER modules on the financial statements of Deutsche Bank from 2021. For this, see the file available on Moodle.

1. Download it and convert the PDF document into text, by using the `pdftotext` command-line utility. In particular, run with the `-layout` option enabled.

In [ ]:
%%bash
. ~/.bashrc
## YOUR SHELL COMMAND HERE
# If you have to execute this command through your shell, still paste the command you ran in here.

2. Given that the document is extremely long, split the inputs into chunks of 500.000 characters and process them separately.

In [ ]:
def load_long_text_in_chunks(fp: str, chunk_size: int = 500_000) -> List[str]:
    """
    Loads a text file (located at `fp`) and chunks it into chunks fo at most `chunk_size` characters.
    Note that the last chunk might be significantly shorter.
    """
    # Load the text file
    text = ## YOUR CODE

    # Split the text into segments of at most `chunk_size` characters
    chunks = ## YOUR CODE
    return chunks

In [ ]:
db_chunks = load_long_text_in_chunks( ## YOUR CODE HERE)

3. Print the top 5 occurring `ORG` entities that are not referencing Deutsche Bank itself, both by using spaCy's NER module and the NER function of NLTK.  
To exclude "Deutsche Bank" entities, filter out all entities that contain both "deutsche" and "bank" in their name, irrespective of the actual upper-/lowercasing.
**Hint:** For more information on how to run NER with NLTK, see [here](https://nanonets.com/blog/named-entity-recognition-with-nltk-and-spacy/#performing-ner-with-nltk-and-spacy)

In [ ]:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('maxent_ne_chunker')
nltk.download('words')

org_entities_spacy = []
org_entities_nltk = []

def is_deutsche_bank_entity(name: str) -> bool:
    """
    Returns True if the entity name contains "deutsche" and "bank" in some upper-/lowercased version.
    This means both "Deutsche Bank" and "deutsche bank's" should be recognized.
    """
    ## YOUR CODE

for chunk in db_chunks:
    # Process the chunk with spaCy
    doc = ## YOUR CODE

    # And also with NLTK
    ## YOUR CODE
    

    # Add all the extracted "ORG" entities to `org_entities`, except those referencing Deutsche Bank
    org_entities_spacy.extend( ## YOUR CODE
    org_entities_nltk.extend( ## YOUR CODE
    

In [ ]:
# Return the top 5 entities by frequency

entity_counts_spacy = ## YOUR CODE
entity_counts_nltk = ## YOUR CODE

print( ## YOUR CODE
print( ## YOUR CODE

4. Compare and analyze the different results between the two methods.

YOUR ANSWER HERE

### Sub Task 3: Co-Occurrence Counts of Entities (5 Points)

As is becoming apparent, the *raw* occurrence counts of entities might not be meaningful on its own, especially if we are interested in less frequently occurring entities.

Instead, we will "investigate" the entities that are most frequently mentioned in association with "Deutsche Bank". For this purpose, we will look at the textual co-occurrences of two named entities. The basic idea is that entities that frequently appear together are likely related.

1. For each text chunk, extract all mentions of the entity `('Deutsche Bank', 'ORG')`, as well as all `PERSON` entity mentions in the text using spaCy. Store the respective entity name and the text position. Unlike the previous question, you do *not* need to check for different spelllings of the "Deutsche Bank" entity.  
**Hint:** Entities are represented as a [`Span`](https://spacy.io/api/span) element in spaCy, which has access to text position.


In [ ]:
entity_mentions_with_start_position = []

for chunk in db_chunks:
    chunk_mentions = []
    # Process the doc with spaCy
    doc = ## YOUR CODE
    
    # Extract only entity mentions of "Deutsche Bank" (ORG) or any PERSON mention.
    # Append each mention, including the text and its starting position, to `chunk_mentions`
    
    # Append the chunk's entities to the aggregate list
    entity_mentions_with_start_position.append(chunk_mentions)


2. Within each chunk, for each mention of `Deutsche Bank`, search for `PERSON` entities that have a starting position within 200 characters before/after the starting position of the `Deutsche Bank` mention. Count for each `PERSON` entity how many times it occurs nearby a mention of `Deutsche Bank`.  
Aggregate the co-occurrences across all chunks. 

In [ ]:
co_occurrences = []

for chunk_mentions in entity_mentions_with_start_position:
    # Iterate through the entities. If the entity is a "Deutsche Bank" mention, extract nearby
    # PERSON references (less than +/- 200 character difference in the starting position)

    ## YOUR CODE


3. Return the number of co-occurrences and the name of the top 5 frequently occurring `PERSON` entities.


In [ ]:
co_occurrence_counts = ## YOUR CODE

print( ## YOUR CODE

4. Look back at the results of your previous task. Are the `PERSON` entities returned by your co-occurrence method the same ones that appear most frequently by raw counts?

YOUR ANSWER HERE

## 3. Neural Models with Huggingface (3 + 5 + 2 = 10 Points)

For state-of-the-art performance, most text-related tasks nowadays use some variation of the Transformer architecture. The particular advantage is especiall the readily available weights for models that have been pre-trained on large general-purpose datasets, which reduces the amount of domain-specific labeled training data.

In this task, we will explore the [Huggingface](https://hf.co/) ecosystem to see in which way Transformer models can be used.
One of the central aspects of the Huggingface platform is the so-called [Model Hub](https://huggingface.co/models), where you can find many different models uploaded by community members for a variety of tasks.

Because the neural models are generally very expensive to run, this exercise will be limited to  less data than in previous questions.

### Sub Task 1: Loading Transformer Models (3 Points)

1. Install the `transformers` library and load the model `cardiffnlp/twitter-roberta-base-sentiment-latest` to classify a sequence.
2. Report the result of the prediction on the test sequence.

In [ ]:
from transformers import ## YOUR IMPORTS

model = ## YOUR CODE
tokenizer = ## YOUR CODE

input_text = "Das ist ein Test."

prediction = ## YOUR CODE

### Sub Task 2: Using Pipelines (5 Points)

The most succinct way of using a Transformer model is the [`transformers.pipeline`](https://huggingface.co/docs/transformers/pipeline_tutorial). You can check out the linked tutorial for more information on the topic, but essentially, `pipeline` provides a light-weight wrapper around a number of different popular NLP tasks

1. Instead of manually defining a pipeline, now load a model through a `"text-classification"` pipeline. Look up the neural model that is loaded by default, and post the link to its [model card](https://huggingface.co/docs/hub/model-cards) below.


In [ ]:
## YOUR CODE

2. Now, instead, load a pipeline for `"text-classification"`, but with a custom model and tokenizer. Use the Model Hub platform to find the most popular model for the German language (by number of downloads) and manually specify the usage of another model (and tokenizer) to the pipeline. Re-run the previous example, and report the prediction result.


In [ ]:

model = ## YOUR CODE
tokenizer = ## YOUR CODE

# Instantiate the pipeline with custom components
pipe = ## YOUR CODE

# Output the prediction by your pipe on the test sample.
print( ## YOUR CODE

3. Keeping in line with the previous exercises, let us now try and actually predict something with the model. Re-load a pipeline, this time for Named Entity Recognition, using the default model.

In [ ]:
## YOUR CODE

4. Run the pipeline with the text from the Deutsche Bank report from Question 2 and output the results.

In [ ]:
## YOUR CODE

print( ## YOUR CODE

5. Look at the results. Something looks strange here; why is it not working properly? Elaborate your answer.

YOUR ANSWER HERE

### Sub Task 3: Using Datasets through Huggingface (2 Points)

Instead of using the `transformers` library for model training and inference, it is also possible to use other libraries by Huggingface without neural models.
In particular, the `datasets` library provides a centralized and streamlined way of accessing a variety of different datasets.

1. Using the `datasets` library, load the `imdb` dataset.

In [ ]:
## YOUR CODE

2. Report the mean length of `text` column for the training, validation and test split, respectively.


In [ ]:
## YOUR CODE